### PR TITLE
Api reorganization

### PR DESCRIPTION
**Pull request by mvadari**
Move all API URLs in the frontend to a separate file, so you don't have to go hunting for the URLs to edit them
Change the API URL hierarchy to follow standard REST API practice ([source](https://medium.com/hashmapinc/rest-good-practices-for-api-design-881439796dc9)) ([hierarchy](https://docs.google.com/document/d/1lgPza4EP7eM_kbxA9F6Uxt7tgZgLjZsbVQwhNVYJUS4/edit))
Rename/create more permissions (some of these aren't really used because I can't figure out how)

### CODE REVIEW

File: server/core/views/api.py [hunk 1]
```diff
@@ -207,6 +198,14 @@ def post(self, request, format=None):
         # TODO add correction checks
         return Response({"message": "Success!"})
 
+    def get_object(self, program=None, edition=None):
```
Comment (evantey14) [inline server/core/views/api.py:201]: why are `program` and `edition` defaulted to `None`?

also is there a particular reason this is named `get_object`? if not, I think it might be nice to have a more specific name that better describes what we're getting

File: server/core/views/api.py [hunk 2]
```diff
@@ -215,23 +214,29 @@ class MedicalLiability(APIView):
...
 
         # TODO add correction checks
         return Response({"message": "Success!"})
 
+    def get_object(self, program=None, edition=None):
+        user = self.request.user
+        program = Program.objects.get(name=program, edition=edition)
+        # TODO add check for if studentreg object exists
+        studentreg = StudentRegistration.objects.get(student=user, program=program)
+        self.check_object_permissions(self.request, studentreg)
```
Comment (evantey14) [inline server/core/views/api.py:236]: why is this line necessary? (particularly if it wasn't necessary in emergency info?

File: server/core/views/api.py [hunk 3]
```diff
@@ -215,23 +214,29 @@ class MedicalLiability(APIView):
...
         # TODO maybe add other permissions to this so you can't accidentally do this?
+        # Like something from formstack
 
         studentreg.medliab_check = True
         studentreg.save()
 
         # TODO add correction checks
         return Response({"message": "Success!"})
 
+    def get_object(self, program=None, edition=None):
```
Comment (evantey14) [inline server/core/views/api.py:231]: same comments as in emergency info

File: server/core/views/api.py [hunk 4]
```diff
@@ -257,6 +258,15 @@ def post(self, request, format=None):
...
```
Comment (evantey14) [inline server/core/views/api.py:261]: same comment as above --- honestly, there's probably a better place to put this code so it's not written 3 times

File: server/core/views/api.py [hunk 5]
```diff
@@ -265,23 +275,27 @@ class Availability(APIView):
...
+    def post(self, request, program, edition, format=None):
+        studentreg = self.get_object(program, edition)
 
         studentreg.availability_check = True
         studentreg.save()
 
         # TODO add correction checks
         return Response({"message": "Success!"})
 
+    def get_object(self, program=None, edition=None):
```
Comment (evantey14) [inline server/core/views/api.py:291]: same comments as above

File: server/core/urls.py [hunk 6]
```diff
@@ -6,7 +6,7 @@
 from . import views
 
 router = routers.DefaultRouter()
-router.register("teacherprograms", views.TeacherProgramViewSet)
+router.register("dashboard/teacher/", views.TeacherProgramViewSet)
```
[the remark this answers is not in the document]
Comment (mvadari) [reply server/core/urls.py:9]: Viewsets work differently than views. I think viewsets can be made to work the same, but it's a little more complicated to get them to do the thing, so it was easier to just do it separately. It'll be less weird once all the other Django URLs are gone.

File: server/core/views/api.py (hunk 1 above)
Comment (mvadari) [reply server/core/views/api.py:201]: It's overriding a method, and [appears to be common practice](https://www.django-rest-framework.org/tutorial/3-class-based-views/)

File: server/core/views/api.py (hunk 2 above)
Comment (mvadari) [reply server/core/views/api.py:236]: I don't quite understand how object permissions work, I need to look more into that. I'll take out the line and add a TODO.
